src/Indigo/Contracts/HomebaseLite/Optimizer.hs

```diff
@@ -0,0 +1,222 @@
+-- SPDX-FileCopyrightText: 2022 Tezos Commons
+-- SPDX-License-Identifier: LicenseRef-MIT-TC
+
+module Indigo.Contracts.HomebaseLite.Optimizer
+    ( optimize
+    ) where
+
+import Indigo (ContractCode, optimizeLorentzWithConf)
+
+import Data.Default (def)
+import Data.Type.Equality
+import Unsafe.Coerce
+
+import Morley.Michelson.Optimizer hiding (optimize)
+import Morley.Michelson.Typed.Instr
+import Morley.Michelson.Typed.T
+import Morley.Michelson.Typed hiding (ContractCode, S)
+import Morley.Util.PeanoNatural
+import Morley.Util.Peano
+import Morley.Util.Type (type (++))
+
+optimize :: ContractCode ps st -> ContractCode ps st
+optimize = optimizeLorentzWithConf optimizerConf
+
+-- | Turn rule fixpoint into rule.
+fixpoint :: (Rule -> Rule) -> Rule
+fixpoint r = go
+  where
+    go :: Rule
+    go = whileApplies (r go)
+
+-- | Apply a rule to the same code, until it fails.
+whileApplies :: Rule -> Rule
+whileApplies r = Rule go
+  where
+    go :: Instr inp out -> Maybe (Instr inp out)
+    go i = maybe (Just i) go (unRule r $ i)
+
+flattenSeqLHS :: Rule -> Rule
+flattenSeqLHS toplevel = Rule \case
+  it@(Seq (Seq _ _) _) -> Just $ linearizeAndReapply toplevel it
+  _                    -> Nothing
+
+linearizeAndReapply :: Rule -> Instr inp out -> Instr inp out
+linearizeAndReapply restart = \case
+  Seq (Seq a b) c ->
+    applyOnce restart $ Seq a (linearizeAndReapply restart (Seq b c))
+
+  other -> applyOnce restart other
+
+applyOnce :: Rule -> Instr inp out -> Instr inp out
+applyOnce r i = maybe i id (unRule r $ i)
+
+optimizerConf :: OptimizerConf
+optimizerConf = def {
+  ocRuleset = ocRuleset def <> fmap fixpoint
+    (seqAdapter (flattenSeqLHS
+      `orSimpleRule` unrollDrops
+      `orSimpleRule` rollDup
+      `orSimpleRule` rollDip)
+    : replicate 10 mainStage
+    <>
+     [seqAdapter $ flattenSeqLHS
+      `orSimpleRule` simpleSynonyms
+      `orSimpleRule` rollDrops
+     ]
+    ) <> ocRuleset def
+  }
+  where
+    mainStage =
+      seqAdapter (flattenSeqLHS
+        `orSimpleRule` dupDipDrop
+        `orSimpleRule` pushDipDug
+        `orSimpleRule` dupDugDrop
+        `orSimpleRule` notIf
+        `orSimpleRule` dupDipDownstream
+        `orSimpleRule` digDug
+        `orSimpleRule` redundantIf
+        `orSimpleRule` nestedDip
+        `orSimpleRule` emptyDip
+        `orSimpleRule` dropFrame
+        ) `orSimpleRule` dropNop
+
+dropFrame :: Rule
+dropFrame = Rule \case
+  FrameInstr Proxy x :# xs -> Just . unsafeCoerce $ x :# unsafeCoerce xs
+  _ -> Nothing
+
+dropNop :: Rule
+dropNop = Rule \case
+  x :# Nop -> Just x
+  Nop :# xs -> Just xs
+  _ -> Nothing
+
+rollDip :: Rule
+rollDip = Rule \case
+  DIP x :# xs -> Just $ DIPN One x :# xs
+  _ -> Nothing
+
+redundantIf :: Rule
```

Comment:
Ok, I see that _a few_ of these rules don't use `unsafeCoerce`, can you please open an MR to upstream these to `morley`?

Reply:
@lierdakil did you miss this or am I missing something?

Reply:
Neither, didn't get to it yet. Actually, this particular rule, `redundantIf`, is in most cases pretty useless. But doesn't hurt to have it, probably.

Reply:
> `redundantIf`, is in most cases pretty useless. But doesn't hurt to have it, probably.

Indeed, probably won't be very useful taken on its own (although mistakes do happen), but might be a bit more when combined with other rules, after these are applied.